mindmeld/models/nn_utils/input_encoders.py

```diff
@@ -0,0 +1,937 @@
+            # If padding_length is None, padding has to be done to the max length of the input batch
+            padding_length = min(max_curr_len, padding_length) if padding_length else max_curr_len
+
+            _trim_length = padding_length - 2  # -2 to sum upto padding_length after adding cls, sep
+            if labels:
+                trimmed_tokenized_examples, trimmed_labels = zip(*[
+                    self._trim_combined(t_ex, _trim_length, _labels)
+                    for t_ex, _labels in zip(tokenized_examples, labels)
+                ])
+            else:
+                trimmed_tokenized_examples = [
+                    self._trim_combined(t_ex, _trim_length) for t_ex in tokenized_examples
+                ]
+            split_lengths = [[len(x) for x in ex] for ex in trimmed_tokenized_examples]
+            trimmed_tokenized_examples = [" ".join(sum(ex, [])) for ex in
+                                          trimmed_tokenized_examples]
+            _inputs = self._bert_tokenizer(
+                trimmed_tokenized_examples, padding=True, truncation=True, max_length=None,
+                return_tensors="pt"
+            ).to(self.device)
+            # last_hidden_state, pooler_output, seq_lengths are the keys outputted
+            _outputs = self._bert_model(**_inputs, return_dict=True)
+
+            # discard [CLS] token, [SEP] token is anyway discarded as the sum of each split length
+            # item does not go as far as the SEP token
+            last_hidden_state = _outputs["last_hidden_state"][:, 1:]
+
+            return_dict.update({
+                "last_hidden_state": last_hidden_state,  # [BS, SEQ_LEN, EMD_DIM]
+                "split_lengths": split_lengths,  # List[List[Int]]
+                "seq_lengths": torch.as_tensor(
+                    [len(_split_lengths) for _split_lengths in split_lengths], dtype=torch.long
+                )
+            })
+
+            if labels:
+                trimmed_labels = pad_sequence(
+                    [torch.as_tensor(label) for label in trimmed_labels], batch_first=True,
+                    padding_value=LABEL_PAD_TOKEN_IDX
+                ).long()
+                return_dict.update({"labels": trimmed_labels})
+
+        return return_dict
+
+
+# sub-classes that define more complex custom `batch_encode` on top of derived encoders
+
+
+class TokenClsDualEncoderForEmbLayers(TokenClsEncoderForEmbLayer):
+    """Dual encoder that encodes both word level as well as character level tokens
+    """
+
+    ALLOWED_TOKENIZERS = [None, "whitespace-tokenizer"]
+    BASIC_SPECIAL_CHAR_VOCAB_DICT = {
+        "char_pad_token": "<CHAR_PAD>",
+        "char_unk_token": "<CHAR_UNK>",
+        "char_start_token": "<CHAR_START>",
+        "char_end_token": "<CHAR_END>",
+    }
+    DEFAULT_PARAMS = {
+        "emb_dim": 300,
+        "char_emb_dim": 50
+    }
+
+    def fit(
```

Comment:
I think the fit method should be named something else here. its not really doing a fit in a ML sense

Reply:
After some refactors, felt `prepare` to be a suitable name for input encoders. Modified to it from `fit`